src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/ElasticSearchShareLogDAO.scala

```diff
@@ -0,0 +1,133 @@
+package org.broadinstitute.dsde.firecloud.dataaccess
+
+import java.time.Instant
+
+import org.broadinstitute.dsde.firecloud.FireCloudException
+import org.broadinstitute.dsde.firecloud.model.ModelJsonProtocol.ShareFormat
+import org.broadinstitute.dsde.firecloud.model.ShareLog.Share
+import org.broadinstitute.dsde.workbench.util.health.SubsystemStatus
+import org.elasticsearch.action.admin.indices.create.{CreateIndexRequest, CreateIndexRequestBuilder, CreateIndexResponse}
+import org.elasticsearch.action.admin.indices.exists.indices.{IndicesExistsRequest, IndicesExistsRequestBuilder, IndicesExistsResponse}
+import org.elasticsearch.action.get.{GetRequest, GetRequestBuilder, GetResponse}
+import org.elasticsearch.action.index.{IndexRequest, IndexRequestBuilder, IndexResponse}
+import org.elasticsearch.action.search.{SearchRequest, SearchRequestBuilder, SearchResponse}
+import org.elasticsearch.action.support.WriteRequest.RefreshPolicy
+import org.elasticsearch.client.transport.TransportClient
+import org.elasticsearch.common.xcontent.XContentType
+import org.elasticsearch.index.query.QueryBuilder
+import org.elasticsearch.index.query.QueryBuilders._
+import spray.json._
+import spray.json.DefaultJsonProtocol._
+
+import scala.concurrent.Future
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.util.hashing.MurmurHash3
+import scala.util.{Failure, Success, Try}
+
+trait ShareQueries {
+  def userShares(userId: String): QueryBuilder = termQuery("userId", userId)
+  def userSharesOfType(userId: String, shareType: String) = boolQuery()
+    .must(termQuery("userId", userId))
+    .must(termQuery("shareType", shareType))
```

Comment:
I think you can reuse `userShares` here:
```
def userSharesOfType(userId: String, shareType: String) = boolQuery()
    .must(userShares(userId))
    .must(termQuery("shareType", shareType))
```

also, a bit of a query optimization - we probably want to use filter context here, not query context.
```
    boolQuery().filter(boolQuery()
    .must(termQuery("userId", userId))
    .must(termQuery("shareType", shareType)))
```
see https://www.elastic.co/guide/en/elasticsearch/reference/5.4/term-level-queries.html for an explanation of this - in short, filter context turns off a bunch of relevance calculations and other search features that we don't need, because these are exact-match queries.

Reply:
Cool, do you think these relevance calculations will come in handy when keeping track of times shared?